packages/driver/src/cy/testConfigOverrides.ts

```diff
@@ -1,17 +1,72 @@
 import _ from 'lodash'
+import $errUtils from '../cypress/error_utils'
+
+// See Test Config Overrides in ../../../../cli/types/cypress.d.ts
+
+type ResolvedTestConfigOverride = {
+  /**
+   * The list of test config overrides and the invocation details used to add helpful
+   * error messaging to consumers if a test override fails validation.
+   */
+  testConfigList: Array<TestConfig>
+  /**
+   * The test config overrides that will apply to the test if it passes validation.
+   * */
+  unverifiedTestConfig: Object
+}
 
-function mutateConfiguration (testConfigOverride, config, env) {
-  const globalConfig = _.clone(config())
-  const globalEnv = _.clone(env())
+type TestConfig = {
+  overrides: {
+    browser?: Object
+  }
+  invocationDetails: {
+    stack: Object
+  }
+};
+
+type ConfigOverrides = {
+  env: Object | undefined
+};
+
+function setConfig (testConfigList: Array<TestConfig>, config, localConfigOverrides: ConfigOverrides = { env: undefined }) {
+  testConfigList.forEach(({ overrides: testConfigOverride, invocationDetails }) => {
+    if (_.isArray(testConfigOverride)) {
+      setConfig(testConfigOverride, config, localConfigOverrides)
+    } else {
+      delete testConfigOverride.browser
+
+      try {
+        config(testConfigOverride)
+      } catch (e) {
+        let err = $errUtils.errByPath('config.invalid_test_override', {
+          errMsg: e.message,
+        })
+
+        err.stack = $errUtils.stackWithReplacedProps({ stack: invocationDetails.stack }, err)
+        throw err
```

Comment:
Just wrote a [better abstraction for this](https://github.com/cypress-io/cypress/pull/18587/files#diff-8ca2c866dca492abea845c9ef35963f7a539e0ad761cdca8b0ac0a1e408e27b0R135) in David's PR. Don't need to change it in this PR, since that would mean waiting for his to get merged first, but I wanted to note that this could be cleaned up in the future.

Reply:
nice! Will be nice to clean this up later on.